app/models/terminal_executor.rb

```diff
@@ -22,43 +22,43 @@ class TerminalExecutor
 
   attr_reader :pid, :pgid, :output, :timeout
 
-  def initialize(output, project:, verbose: false, deploy: nil)
+  def initialize(output, project:, verbose: false, deploy: nil, cancel_timeout: 1, timeout: 5)
     @output = output
     @verbose = verbose
     @deploy = deploy
     @project = project
-    @cancelled = false
-    @timeout = Rails.application.config.samson.deploy_timeout
+    @timeout = timeout
+    @cancel_timeout = cancel_timeout
   end
 
   def execute(*commands, timeout: @timeout)
-    return false if @cancelled
-    options = {in: '/dev/null', unsetenv_others: true}
     script_as_executable(script(commands)) do |command|
-      output, input, pid = PTY.spawn(whitelisted_env, command, options)
-      record_pid(pid) do
-        timeout_execution(timeout) do
-          stream from: output, to: @output
+      output, input, pid = PTY.spawn(whitelisted_env, command, in: '/dev/null', unsetenv_others: true)
+      input.close # close early so nothing can hang forever
```

Comment:
this is new - have you tested this doesn't break weird streaming / process behaviour?